### PR TITLE
FUSETOOLS2-395 - added a Delete VDB command for the project tree node

### PR DESCRIPTION
**Pull request by lhein**
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

### CODE REVIEW

File: src/test/commands.vdb.test.ts [hunk 1]
```diff
@@ -78,7 +79,42 @@ describe('Commands Tests', () => {
...
+			should.equal(true, success, `Execution of the deleteVDBCommand returned false for VDB ${name} with file ${vdbFile}`);
+			fs.existsSync(vdbFile).should.equal(false);
+			const dvConfig = utils.loadModelFromFile(vdbFile);
+			should.not.exist(dvConfig);
+		});
+
+		it('should not delete a VDB file when handing over invalid file name', async () => {
+			try {
+				await deleteVDBCommand.handleVDBDeletion(name, `${vdbFile}.not.existing.yaml`);
+				fail('delete command did not throw exception when handing over invalid file');
```
Comment (bfitzpat) [inline src/test/commands.vdb.test.ts:108]: Same -- "handing over" vs. "passing"?

File: src/test/commands.vdb.test.ts [hunk 2]
```diff
@@ -78,7 +79,42 @@ describe('Commands Tests', () => {
...
+			} catch (error) {
+				should.exist(error);
+			}
+		});
+
+		it('should not generate a VDB file when handing over undefined as file', async () => {
+			try {
+				await createVDBCommand.handleVDBCreation(name, undefined);
+				fail('delete command did not throw exception when handing over undefined instead of the file');
```
Comment (bfitzpat) [inline src/test/commands.vdb.test.ts:117]: Same -- "handing over" vs. "passing"?

File: src/commands/DeleteVDBCommand.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import * as fs from 'fs';
+import * as vscode from 'vscode';
+import { DVProjectTreeNode } from '../model/tree/DVProjectTreeNode';
+
+export function deleteVDBCommand(dvProjectNode: DVProjectTreeNode): void {
+	if (dvProjectNode) {
+		handleVDBDeletion(dvProjectNode.label, dvProjectNode.file)
+			.then( (success: boolean) => {
+				if (success) {
+					vscode.window.showInformationMessage(`VDB ${dvProjectNode.label} has been deleted...`);
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/commands/DeleteVDBCommand.ts:27]: I agree. Would be better to be consistent in naming.

File: src/test/commands.vdb.test.ts [hunk 4]
```diff
@@ -69,7 +70,7 @@ describe('Commands Tests', () => {
 		it('should not generate a VDB file when handing over invalid file name', async () => {
 			try {
 				await createVDBCommand.handleVDBCreation(workspacePath, undefined, templateFolder);
-				fail('create command did not throw exception when handing invalid name');
+				fail('create command did not throw exception when handing over invalid name');
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/test/commands.vdb.test.ts:73]: Its just a test. Only developers will see those messages and can easily do a "Find" on that.

File: package.json [hunk 5]
```diff
@@ -124,6 +129,10 @@
 					"command": "datavirt.delete.datasource",
 					"when": "false"
 				},
+				{
+					"command": "datavirt.delete.datasource",
+					"when": "false"
+				},
```
Comment (apupier) [inline package.json:135]: this looks a like a wrong copy. The command is duplicated

File: src/commands/DeleteVDBCommand.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+				if (success) {
+					vscode.window.showInformationMessage(`Virtual database ${dvProjectNode.label} has been deleted...`);
+				} else {
+					vscode.window.showErrorMessage(`An error occured when trying to delete the virtual database ${dvProjectNode.label}...`);
+				}
+		});
+	}
+}
+
+export function handleVDBDeletion(vdbName: string, file: string): Promise<boolean> {
```
Comment (apupier) [inline src/commands/DeleteVDBCommand.ts:35]: I don't see the benefit of returning a Promise here. Am I missing something?

what about using boolean directly?

```
export function deleteVDBCommand(dvProjectNode: DVProjectTreeNode): void {
	if (dvProjectNode) {
		const success = handleVDBDeletion(dvProjectNode.label, dvProjectNode.file);
		if (success) {
			vscode.window.showInformationMessage(`Virtual database ${dvProjectNode.label} has been deleted...`);
		} else {
			vscode.window.showErrorMessage(`An error occured when trying to delete the virtual database ${dvProjectNode.label}...`);
		}
	}
}

export function handleVDBDeletion(vdbName: string, file: string): boolean {
	if (vdbName && file) {
		try {
			fs.unlinkSync(file);
			return true;
		} catch (error) {
			extension.log(error);
		}
	} else {
		extension.log(`handleVDBDeletion: Unable to delete the virtual database ${vdbName}...`);
	}
	return false;
}
```

File: src/test/commands.vdb.test.ts [hunk 7]
```diff
@@ -66,19 +67,54 @@ describe('Commands Tests', () => {
...
+			}
+		});
+
+		it('should not generate a VDB file when passing undefined as file', async () => {
+			try {
+				await createVDBCommand.handleVDBCreation(name, undefined);
+				fail('delete command did not throw exception when passing undefined instead of the file');
 			} catch (error) {
 				should.exist(error);
 			}
```
Comment (apupier) [inline src/test/commands.vdb.test.ts:120]: looks alike this test is abput creation, it has moved in the "delete" test context.

File: package.json (hunk 5 above)
Comment (lhein) [reply package.json:135]: Ouch...that is a copy/paste error. it should be datavirt.delete.vdb.

File: src/commands/DeleteVDBCommand.ts (hunk 6 above)
Comment (lhein) [reply src/commands/DeleteVDBCommand.ts:35]: Good catch. Will check that. It might be the case for several other commands too.

Comment (lhein) [reply src/commands/DeleteVDBCommand.ts:35]: Actually using a Promise is a good thing here. The real issue is that I use a synchronous file system operation which basically blocks the whole browser / vscode ui. I will provide a fix soon and also raise a Jira to also fix other places where I invoke synchronous fs operations.

**Comment by lhein** [comment]
I added a logic to close the editor for the yaml file if it is open when deleting the VDB. Actually I will try to find a command of VS Code to do the delete. I think it might also contain a logic for the closing of editors. Keeping this open until I did the research.

**Comment by lhein** [comment]
Found out that vscode has its own api layer for the file system with `vscode.workspace.fs` which seems to be the more consistent way to manipulate the file system. The delete now behaves the same as the "Delete" command for the files in the explorer view and I was able to get rid off the external fs and utils calls.

File: src/commands/DeleteVDBCommand.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+		vscode.window.showInformationMessage(`Virtual database ${dvProjectNode.label} has been deleted...`);
+	} else {
+		vscode.window.showErrorMessage(`An error occured when trying to delete the virtual database ${dvProjectNode.label}...`);
+	}
+}
+
+export async function handleVDBDeletion(vdbName: string, file: string): Promise<boolean> {
+	if (vdbName && file) {
+		try {
+			await vscode.workspace.fs.delete(vscode.Uri.parse(file));
```
Comment (apupier) [inline src/commands/DeleteVDBCommand.ts:33]: as the pathname doesn't contain the "file://",  vscode.Uri.file(file) should be used instead. it allows to handle special characters such as `#` and `?` that are part of the path of the file

to reproduce:
- open a folder for VS Code instance using a # in the path
- create a new Datavirtualization definition file
- right-click to delete it

**Comment by lhein** [comment]
> using `vscode.workspace.fs` is so much better! Feel dumb that I missed this API before.

Yeah, same here.

File: src/commands/DeleteVDBCommand.ts (hunk 8 above)
Comment (lhein) [reply src/commands/DeleteVDBCommand.ts:33]: I followed the steps and it worked fine. Wondering why you have an issue.

Comment (lhein) [reply src/commands/DeleteVDBCommand.ts:33]: Oh wait. I used file delete and not the Delete Virtual Database. Going to fix this now.

**Pull request merged**